### DISCUSSION TITLE
Help on ESP32 build openthread

### DISCUSSION THREAD

**Discussion by RolliGrolli** [openthread/openthread#7799] [Q&A] [answered]
I am using the ESP32 with the openthread border router project (esp-idf/examples/openthread/ot_br/). I would like to make some changes inside the openthread part. The openthread functionality is encapsulated inside the libopenthread_port.a / libopentherad_br.a libraries under /esp-idf/components/opentherad/lib/esp32. To let the compiler build the library i tried to enable the CONFIG_OPENTHREAD_ESP_LIB_FROM_INTERNAL_SRC inside the menuconfig. But if try then to compile the project, it fails with: 

Executing action: all (aliases: build)
Running ninja in directory /home/groll/tools/esp-idf/examples/openthread/ot_br/build
Executing "ninja all"...
[0/1] Re-running CMake...
-- Building ESP-IDF components for target esp32
-- Project sdkconfig file /home/groll/tools/esp-idf/examples/openthread/ot_br/sdkconfig
Loading defaults file /home/groll/tools/esp-idf/examples/openthread/ot_br/sdkconfig.defaults...
-- App "esp_ot_br" version: v4.4-dirty
-- Adding linker script /home/groll/tools/esp-idf/examples/openthread/ot_br/build/esp-idf/esp_system/ld/memory.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_system/ld/esp32/sections.ld.in
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.api.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.libgcc.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.newlib-data.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.syscalls.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld
-- Adding linker script /home/groll/tools/esp-idf/components/esp_rom/esp32/ld/esp32.rom.newlib-time.ld
CMake Error at ../../../tools/cmake/component.cmake:369 (__component_get_property):
  __component_get_property Function invoked with incorrect arguments for
  function named: __component_get_property
Call Stack (most recent call first):
  ../../../components/openthread/CMakeLists.txt:52 (idf_component_get_property)

CMake Error at ../../../tools/cmake/component.cmake:369 (__component_get_property):
  __component_get_property Function invoked with incorrect arguments for
  function named: __component_get_property
Call Stack (most recent call first):
  ../../../components/openthread/CMakeLists.txt:59 (idf_component_get_property)

-- Adding linker script /home/groll/tools/esp-idf/components/soc/esp32/ld/esp32.peripherals.ld
-- Components: app_trace app_update asio bootloader bootloader_support bt cbor cmock coap console cxx driver efuse esp-tls esp32 esp_adc_cal esp_common esp_eth esp_event esp_gdbstub esp_hid esp_http_client esp_http_server esp_https_ota esp_https_server esp_hw_support esp_ipc esp_lcd esp_local_ctrl esp_netif esp_phy esp_pm esp_ringbuf esp_rom esp_serial_slave_link esp_system esp_timer esp_websocket_client esp_wifi espcoredump esptool_py expat fatfs freemodbus freertos hal heap idf_test ieee802154 jsmn json libsodium log lwip main mbedtls mdns mqtt newlib nghttp nvs_flash openssl openthread partition_table perfmon protobuf-c protocol_examples_common protocomm pthread sdmmc soc spi_flash spiffs tcp_transport tcpip_adapter tinyusb ulp unity usb vfs wear_levelling wifi_provisioning wpa_supplicant xtensa
-- Component paths: /home/groll/tools/esp-idf/components/app_trace /home/groll/tools/esp-idf/components/app_update /home/groll/tools/esp-idf/components/asio /home/groll/tools/esp-idf/components/bootloader /home/groll/tools/esp-idf/components/bootloader_support /home/groll/tools/esp-idf/components/bt /home/groll/tools/esp-idf/components/cbor /home/groll/tools/esp-idf/components/cmock /home/groll/tools/esp-idf/components/coap /home/groll/tools/esp-idf/components/console /home/groll/tools/esp-idf/components/cxx /home/groll/tools/esp-idf/components/driver /home/groll/tools/esp-idf/components/efuse /home/groll/tools/esp-idf/components/esp-tls /home/groll/tools/esp-idf/components/esp32 /home/groll/tools/esp-idf/components/esp_adc_cal /home/groll/tools/esp-idf/components/esp_common /home/groll/tools/esp-idf/components/esp_eth /home/groll/tools/esp-idf/components/esp_event /home/groll/tools/esp-idf/components/esp_gdbstub /home/groll/tools/esp-idf/components/esp_hid /home/groll/tools/esp-idf/components/esp_http_client /home/groll/tools/esp-idf/components/esp_http_server /home/groll/tools/esp-idf/components/esp_https_ota /home/groll/tools/esp-idf/components/esp_https_server /home/groll/tools/esp-idf/components/esp_hw_support /home/groll/tools/esp-idf/components/esp_ipc /home/groll/tools/esp-idf/components/esp_lcd /home/groll/tools/esp-idf/components/esp_local_ctrl /home/groll/tools/esp-idf/components/esp_netif /home/groll/tools/esp-idf/components/esp_phy /home/groll/tools/esp-idf/components/esp_pm /home/groll/tools/esp-idf/components/esp_ringbuf /home/groll/tools/esp-idf/components/esp_rom /home/groll/tools/esp-idf/components/esp_serial_slave_link /home/groll/tools/esp-idf/components/esp_system /home/groll/tools/esp-idf/components/esp_timer /home/groll/tools/esp-idf/components/esp_websocket_client /home/groll/tools/esp-idf/components/esp_wifi /home/groll/tools/esp-idf/components/espcoredump /home/groll/tools/esp-idf/components/esptool_py /home/groll/tools/esp-idf/components/expat /home/groll/tools/esp-idf/components/fatfs /home/groll/tools/esp-idf/components/freemodbus /home/groll/tools/esp-idf/components/freertos /home/groll/tools/esp-idf/components/hal /home/groll/tools/esp-idf/components/heap /home/groll/tools/esp-idf/components/idf_test /home/groll/tools/esp-idf/components/ieee802154 /home/groll/tools/esp-idf/components/jsmn /home/groll/tools/esp-idf/components/json /home/groll/tools/esp-idf/components/libsodium /home/groll/tools/esp-idf/components/log /home/groll/tools/esp-idf/components/lwip /home/groll/tools/esp-idf/examples/openthread/ot_br/main /home/groll/tools/esp-idf/components/mbedtls /home/groll/tools/esp-idf/components/mdns /home/groll/tools/esp-idf/components/mqtt /home/groll/tools/esp-idf/components/newlib /home/groll/tools/esp-idf/components/nghttp /home/groll/tools/esp-idf/components/nvs_flash /home/groll/tools/esp-idf/components/openssl /home/groll/tools/esp-idf/components/openthread /home/groll/tools/esp-idf/components/partition_table /home/groll/tools/esp-idf/components/perfmon /home/groll/tools/esp-idf/components/protobuf-c /home/groll/tools/esp-idf/examples/common_components/protocol_examples_common /home/groll/tools/esp-idf/components/protocomm /home/groll/tools/esp-idf/components/pthread /home/groll/tools/esp-idf/components/sdmmc /home/groll/tools/esp-idf/components/soc /home/groll/tools/esp-idf/components/spi_flash /home/groll/tools/esp-idf/components/spiffs /home/groll/tools/esp-idf/components/tcp_transport /home/groll/tools/esp-idf/components/tcpip_adapter /home/groll/tools/esp-idf/components/tinyusb /home/groll/tools/esp-idf/components/ulp /home/groll/tools/esp-idf/components/unity /home/groll/tools/esp-idf/components/usb /home/groll/tools/esp-idf/components/vfs /home/groll/tools/esp-idf/components/wear_levelling /home/groll/tools/esp-idf/components/wifi_provisioning /home/groll/tools/esp-idf/components/wpa_supplicant /home/groll/tools/esp-idf/components/xtensa
-- Configuring incomplete, errors occurred!
See also "/home/groll/tools/esp-idf/examples/openthread/ot_br/build/CMakeFiles/CMakeOutput.log".
FAILED: build.ninja 
/usr/bin/cmake -S/home/groll/tools/esp-idf/examples/openthread/ot_br -B/home/groll/tools/esp-idf/examples/openthread/ot_br/build
ninja: error: rebuilding 'build.ninja': subcommand failed
ninja failed with exit code 1

How can I compile the thread library? Is there another way to do it, maybe standalone compiling? 
Or do I have to do smthg elese?

**Comment by jwhui** [comment]
Maybe @chshu or @gjc13 from Espressif can help provide some insight.

**Comment by chshu** [comment, marked as answer]
@RolliGrolli The current ESP Thread Border Router is released in static library, so you can't build from source code. The `CONFIG_OPENTHREAD_ESP_LIB_FROM_INTERNAL_SRC` option is for the internal use only.

We are more than happy to support you on new feature requests, feel free to raise an issue on https://github.com/espressif/esp-idf/issues. If you want some deep discussion, please contact us in this link: https://www.espressif.com/en/contact-us/technical-inquiries/new-feature-requests.

ESP Thread Border Router SDK and Docs:
https://github.com/espressif/esp-thread-br
https://docs.espressif.com/projects/esp-thread-br/en/latest/esp32/

**Reply by RolliGrolli** [reply]
Is the compilation of the library very difficult? Which toolchain will be used to generate the library?

**Reply by chshu** [reply]
Not related to the toolchain, the main blocker is that the ESP border router source code is not public yet..

**Reply by MartyMcFlyInTheSky** [reply]
@chshu Any timeline when the full thread source code will be public from espressif side?

**Reply by chshu** [reply]
@MartyMcFlyInTheSky No specific timeline yet... Feel free to let us know if any feature requests or bug reports.

**Reply by mvcorrea-ufc** [reply]
The default esp32S3 firmware come with the web interface enabled? How to enable it?
LOTS of posts on ESP32 forum about not being able to compile the S3 firmware!
Only able to reach the cli via the S3 USB.
As I cant build it from source you gonna make next releases alvailable?

**Reply by chshu** [reply]
@mvcorrea-ufc The Web server is not enabled by default. Enable the `CONFIG_OPENTHREAD_BR_START_SERVER` option in menuconfig to enable the Web Server feature.

Please refer to the ESP Thread BR docs page for more details:
- https://docs.espressif.com/projects/esp-thread-br
- https://docs.espressif.com/projects/esp-thread-br/en/latest/esp32/codelab/web-gui.html#prerequisites

The compile issues may caused by some old IDF versions, could you fetch the latest esp-thread-br, and IDF master or v5.1.1 release. Let me know if you have any further questions.

**Reply by chshu** [reply]
@mvcorrea-ufc The Web server is not enabled by default. Enable the `CONFIG_OPENTHREAD_BR_START_SERVER` option in menuconfig to enable the Web Server feature.

Please refer to the ESP Thread BR docs page for more details:
- https://docs.espressif.com/projects/esp-thread-br
- https://docs.espressif.com/projects/esp-thread-br/en/latest/esp32/codelab/web-gui.html#prerequisites

The compile issues may caused by some old IDF versions, could you fetch the latest esp-thread-br, and IDF master or v5.1.1 release. Let me know if you have any further questions.

**Comment by kklem0** [comment]
@chshu the company that I'm working with recommended us to use ESP32 for our product, do you know if ESP border router source code will be released and when?

Thanks!

**Reply by chshu** [reply]
@klem-everywhere The ESP Thread Border Router SDK and Docs are available here:

- https://github.com/espressif/esp-thread-br
- https://docs.espressif.com/projects/esp-thread-br/en/latest/esp32/

Feel free to get in touch with us via: https://www.espressif.com/en/contact-us/sales-questions

**Comment by martin8310** [comment]
Hello,

the file ist always overwritten, so that it can't compile with "CONFIG_OPENTHREAD_BR_START_SERVER=y". 

`CONFIG_OPENTHREAD_BR_AUTO_UPDATE_RCP=y`
`# CONFIG_OPENTHREAD_BR_CREATE_OTA_IMAGE is not set`
`CONFIG_OPENTHREAD_BR_AUTO_START=y`
`CONFIG_OPENTHREAD_BR_START_WEB`

any idea?

regards
Martin

**Reply by chshu** [reply]
You can config it in the IDF kconfig tool, open the configuration panel with command:
```
idf.py menuconfig
```

**Reply by martin8310** [reply]
Thanks, but it doesn't work:


http://192.168.1.22
{
    "ErrorCode": "404",
    "ErrorMessage": "404 Not Found"
}

**Reply by chshu** [reply]
@martin8310 could you file an issue here: https://github.com/espressif/esp-thread-br/issues, with corresponding logs, we will check.
